go-controller/pkg/cluster/ha_master.go

```diff
@@ -0,0 +1,551 @@
+		return nil
+	}
+
+	logrus.Infof("ConfigureAsStandby: New leader IP is : [%s]", masterIPList[0])
+
+	// Get the active ovsdb-server address.
+	var stdout string
+	stdout, _, err = util.RunOVNNBAppCtl("ovsdb-server/get-active-ovsdb-server")
+	if err != nil {
+		logrus.Errorf("Getting  active-ovsdb-server of NB ovsdb-server failed")
+		return err
+	}
+
+	activeServerOutOfSync := false
+	s := strings.Split(stdout, ":")
+	if len(s) != 3 {
+		activeServerOutOfSync = true
+	} else {
+		if s[0] != "tcp" || s[1] != masterIPList[0] || s[2] != strconv.Itoa(int(nbDBPort)) {
+			activeServerOutOfSync = true
+		}
+	}
+
+	if !activeServerOutOfSync {
+		stdout, _, err = util.RunOVNSBAppCtl("ovsdb-server/get-active-ovsdb-server")
+		if err != nil {
+			logrus.Errorf("Getting  active-ovsdb-server of SB ovsdb-server failed")
+			return err
+		}
+
+		s := strings.Split(stdout, ":")
+		if len(s) != 3 {
+			activeServerOutOfSync = true
+		} else {
+			if s[0] != "tcp" || s[1] != masterIPList[0] || s[2] != strconv.Itoa(int(sbDBPort)) {
+				activeServerOutOfSync = true
+			}
+		}
+	}
+
+	if activeServerOutOfSync {
+		logrus.Debugf("ConfigureAsStandby : active server out of sync..Setting the new active server to : " + masterIPList[0])
+		err = hacluster.DemoteOVNDbs(masterIPList[0], nbDBPort, sbDBPort)
+		if err != nil {
+			logrus.Errorf("Demoting OVN ovsdb-servers to standby failed")
+			return err
+		}
+	}
+	return nil
+}
+
+func (hacluster *OvnHAMasterController) validateOvnDbEndpoints(ep *kapi.Endpoints) bool {
+	if ep.Name != ovnkubeDbEp {
+		return false
+	}
+
+	epAnnotations := ep.ObjectMeta.GetAnnotations()
+	leader, present := epAnnotations["ovnkube-master-leader"]
+	if !present || leader != hacluster.leaderElector.GetLeader() {
+		return false
+	}
+
+	masterIPList, sbDBPort, nbDBPort, err := extractDbRemotesFromEndpoint(ep)
+	if err != nil {
+		return false
+	}
+
+	if masterIPList[0] != config.Kubernetes.PodIP ||
+		sbDBPort != ovnSouthDbPort || nbDBPort != ovnNorthDbPort {
+		return false
+	}
+
+	return true
+}
+
+// WatchOvnDbEndpoints watches the ovnkube-db end point
+func (hacluster *OvnHAMasterController) WatchOvnDbEndpoints() error {
+	_, err := hacluster.cluster.watchFactory.AddFilteredEndpointsHandler(config.Kubernetes.OVNConfigNamespace,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				ep := obj.(*kapi.Endpoints)
+				if ep.Name != ovnkubeDbEp {
+					return
+				}
+				if hacluster.leaderElector.IsLeader() {
+					if !hacluster.validateOvnDbEndpoints(ep) {
+						_ = hacluster.updateOvnDbEndpoints(ep, false)
+					}
+				} else {
+					err := hacluster.ConfigureAsStandby(ep)
+					if err != nil {
+						logrus.Errorf(err.Error())
+						panic(err.Error())
+					}
+				}
+			},
+			UpdateFunc: func(old, new interface{}) {
+				epNew := new.(*kapi.Endpoints)
+				if epNew.Name != ovnkubeDbEp {
+					return
+				}
+				if hacluster.leaderElector.IsLeader() {
+					if !validateOVNConfigEndpoint(epNew) {
```

Comment:
@numansiddique don't we want hacluster.validateOvnDbEndpoints() here instead?

Reply:
And acutally, can't Update() just call Add(new)? the code is exactly the same unless I'm missing something.

Reply:
Ack. I will do that.

Reply:
Done